File: command/command/commands/DeleteItem.cs
using command.document;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace command.commands
{
	public class DeleteItem : AbstractCommand
	{
		private readonly IDocument _document;
		private readonly IDocumentItem _item;
		private readonly int _position;

		public DeleteItem(IDocument document, int position)
		{
			_document = document ?? throw new ArgumentNullException(nameof(document));
			if (position > _document.ItemsCount)
			{
				throw new CommandError($"Invalid position {position}");
			}
			_item = _document.GetItem(position);
			_position = position;
		}

		protected override void ExecuteImpl() => _document.DeleteItem(_position);
		protected override void UnexecuteImpl() => _document.InsertItem(_item, _position);

		protected override void DisposeImpl()
		{
			if (IsExecuted)
			{
				_item.Dispose();
			}
		}
	}
}

File: command_java/Document/src/document/commands/DeleteItem.java
package document.commands;

import document.documentItem.ItemsCollection.IDocumentItem;
import document.documentItem.ItemsCollection.IMutableDocumentItemCollection;
import hsitory.AbstractCommand;

import java.util.Optional;

class DeleteItem extends AbstractCommand {

    public DeleteItem(IMutableDocumentItemCollection items, int position) {
        _oldItem = items.GetItem(position);
        _position = position;
        _items = items;
    }

    @Override
    protected void DestroyImpl() throws Exception {
        if (IsExecuted()){
            _oldItem.close();
        }
    }

    @Override
    protected void UnExecuteImpl() {
        _items.InsertItem(_oldItem, _position);
    }

    @Override
    protected void ExecuteImpl() {
        _items.DeleteItem(_position);
    }

    private final int _position;
    private final IDocumentItem _oldItem;
    private final IMutableDocumentItemCollection _items;
}
